File: phi_plugin_js/model/getSave.py
import shutil

from ..models import SstkData
from .cls.common import Save
from .cls.saveHistory import saveHistory
from .getFile import readFile
from .getRksRank import getRksRank
from .path import savePath


class getSave:
    @classmethod
    async def add_user_token(cls, user_id: str, sstk: str):
        """添加 user_id 号对应的 Token"""
        return await SstkData.set_sstk(user_id, sstk)

    @classmethod
    async def get_user_token(cls, user_id: str):
        """获取 user_id 号对应的 Token"""
        return await SstkData.get_sstk(user_id)

    @classmethod
    async def del_user_token(cls, user_id: str):
        """移除 user_id 对应的 Token"""
        return await SstkData.delete_sstk(user_id)

    @classmethod
    async def getSave(cls, user_id: str) -> Save | None:
        """
        获取 user_id 对应的存档文件

        :param str user_id: user_id
        :return: Save
        """
        sstk = await SstkData.get_sstk(user_id)
        if sstk is None:
            return None
        if await cls.isBanSessionToken(sstk):
            raise ValueError(f"{sstk} 已被禁用")

        result = await readFile.FileReader(savePath / sstk / "save.json")
        if not result:
            return None
        tem = await Save().constructor(result)
        if tem.saveInfo:
            await tem.init()
        else:
            return None
        return tem

    @classmethod
    async def getSaveBySessionToken(cls, sstk: str | None) -> Save | None:
        """
        获取 sessionToken 对应的存档文件

        :param str | None sstk: sessionToken
        :return: Save
        """
        if sstk is None:
            return
        if await cls.isBanSessionToken(sstk):
            raise ValueError(f"{sstk} 已被禁用")

        result = await readFile.FileReader(savePath / sstk / "save.json")
        if not result:
            return None
        tem = await Save().constructor(result)
        if tem.saveInfo:
            await tem.init()
        else:
            return None
        return tem

    @classmethod
    async def putSave(cls, user_id: str, data: dict):
        """
        保存 user_id 对应的存档文件

        :param str user_id: user_id
        :param dict data: data
        """
        sstk = data.get("sessionToken")
        if await cls.isBanSessionToken(sstk):
            raise ValueError(f"{sstk} 已被禁用")
        assert sstk is not None
        await cls.add_user_token(user_id, sstk)
        return await readFile.SetFile(savePath / sstk / "save.json", data)

    @classmethod
    async def getHistory(cls, user_id: str) -> "saveHistory":
        """
        获取 user_id 对应的历史记录

        :param str user_id: user_id
        :return: saveHistory
        """
        sstk = await SstkData.get_sstk(user_id)
        if await cls.isBanSessionToken(sstk):
            raise ValueError(f"{sstk} 已被禁用")
        result = (
            await readFile.FileReader(savePath / sstk / "history.json") if sstk else {}
        )
        return saveHistory(result)

    @classmethod
    async def getHistoryBySessionToken(cls, sstk: str) -> "saveHistory":
        if await cls.isBanSessionToken(sstk):
            raise ValueError(f"{sstk} 已被禁用")
        result = await readFile.FileReader(savePath / sstk / "history.json")
        return saveHistory(result)

    @classmethod
    async def putHistory(cls, user_id: str, data: dict):
        """
        保存 user_id 对应的历史记录

        :param str user_id: user_id
        :param dict data: saveHistory
        """
        sstk = await SstkData.get_sstk(user_id)
        if sstk is None:
            return None
        return await readFile.SetFile(savePath / (sstk) / "history.json", data)

    @classmethod
    async def getDan(cls, user_id: str, all=False):
        """
        获取玩家 Dan 数据

        :param str user_id: QQ号
        :param bool all: 是否返回所有数据
        :return: Dan数据
        """
        history = await cls.getHistory(user_id)

        dan = history.dan if history else None

        if dan and not isinstance(dan, list):
            dan = [dan]
        if all is True:
            return dan
        return dan[0] if dan else None

    @classmethod
    async def delSave(cls, user_id: str):
        """
        删除 user_id 对应的存档文件

        :param str user_id: user_id
        """
        sstk = await SstkData.get_sstk(user_id)
        if sstk is None:
            return False
        fPath = savePath / sstk
        await readFile.DelFile(fPath / "save.json")
        await readFile.DelFile(fPath / "history.json")
        await getRksRank.delUserRks(sstk)
        shutil.rmtree(fPath, ignore_errors=True)
        await cls.del_user_token(user_id)
        return True

    @classmethod
    async def delSaveBySessionToken(cls, sstk: str | None):
        """
        删除 sessionToken 对应的存档文件

        :param str sstk: sessionToken
        """
        if sstk is None:
            return False
        fPath = savePath / sstk
        await readFile.DelFile(fPath / "save.json")
        await readFile.DelFile(fPath / "history.json")
        await getRksRank.delUserRks(sstk)
        shutil.rmtree(fPath, ignore_errors=True)
        return True

    @classmethod
    async def banSessionToken(cls, token: str):
        """
        禁用 sessionToken
        :param str token: sessionToken
        """
        return await SstkData.ban_sstk(token)

    @classmethod
    async def allowSessionToken(cls, token: str):
        """
        解禁 sessionToken
        :param str token: sessionToken
        """
        return await SstkData.unban_sstk(token)

    @classmethod
    async def isBanSessionToken(cls, token: str | None):
        """
        检查 sessionToken 是否被禁用
        :param str|None token: sessionToken
        """
        return False if token is None else await SstkData.is_ban_sessionToken(token)

    @classmethod
    async def getGod(cls):
        """
        获取所有被禁用的 sessionToken
        """
        return await SstkData.get_ban_sstk()

File: phi_plugin_koishi/model/getSave.ts
import path from 'path'
import { savePath } from "../components/pluginPath"
import { redis } from '../components/redis'
import getFile from "./getFile"
import Save from './class/Save'
import fs from 'fs'
import saveHistory from './class/saveHistory'
import getRksRank from './getRksRank'
import PhigrosUser from '../lib/PhigrosUser'
// import { redis } from 'yunzai'

export default class getSave {

    /**获取 user_id 号对应的 Token */
    static async get_user_token(user_id: string): Promise<string> {
        return (await redis.get('phigrosUserToken', { userId: user_id }))[0]?.token
    }

    /**添加 user_id 号对应的 Token */
    static async add_user_token(user_id: string, token: string) {
        return await redis.upsert('phigrosUserToken', [{ userId: user_id, token: token }], 'userId')
    }

    /**移除 user_id 对应的 Token */
    static async del_user_token(user_id: string) {
        return await redis.remove('phigrosUserToken', { userId: user_id })
    }

    /**
     * 获取 user_id 对应的存档文件
     * @param {String} user_id user_id
     * @returns {Promise<Save>}
     */
    static async getSave(user_id: string): Promise<Save> {
        let Token = await this.get_user_token(user_id)
        if (await this.isBanSessionToken(Token)) {
            throw new Error(`${Token} 已被禁用`)
        }
        let result = Token ? await getFile.FileReader(path.join(savePath, Token, 'save.json')) : null
        if (result) {
            let tem = new Save(result)
            if (tem.saveInfo) {
                // await tem.init()
            } else {
                return null
            }
            return tem
        } else {
            return null
        }
    }

    /**
     * 获取 sessionToken 对应的存档文件
     * @param {string} Token 
     * @returns 
     */
    static async getSaveBySessionToken(Token: string) {
        // console.info(Token)
        if (await this.isBanSessionToken(Token)) {
            throw new Error(`${Token} 已被禁用`)
        }
        let result = Token ? await getFile.FileReader(path.join(savePath, Token, 'save.json')) : null
        if (result) {
            let tem = new Save(result)
            if (tem.saveInfo) {
                // await tem.init()
            } else {
                return null
            }
            return tem
        } else {
            return null
        }
    }

    /**
     * 保存 user_id 对应的存档文件
     * @param {String} user_id user_id
     * @param {Save} data 
     */
    static async putSave(user_id: string, data: Save | PhigrosUser) {
        let session = data.sessionToken
        if (await this.isBanSessionToken(session)) {
            throw new Error(`${session} 已被禁用`)
        }
        this.add_user_token(user_id, session)
        await getRksRank.addUserRks(session, data.saveInfo.summary.rankingScore)
        return await getFile.SetFile(path.join(savePath, session, 'save.json'), data)
    }

    /**
     * 获取 user_id 对应的历史记录
     * @param {string} user_id 
     * @returns {Promise<saveHistory>}
     */
    static async getHistory(user_id: string): Promise<saveHistory> {
        let session = await this.get_user_token(user_id)
        let result = session ? await getFile.FileReader(path.join(savePath, session, 'history.json')) : null
        return new saveHistory(result)
    }

    /**
     * 保存 user_id 对应的历史记录
     * @param {String} user_id user_id
     * @param {Object} data 
     */
    static async putHistory(user_id: string, data: object) {
        let session = await this.get_user_token(user_id)
        return await getFile.SetFile(path.join(savePath, session, 'history.json'), data)
    }


    /**
     * 获取玩家 Dan 数据
     * @param {string} user_id QQ号
     * @param {boolean} [all=false] 是否返回所有数据
     * @returns {object|Array} Dan数据
     */
    static async getDan(user_id: string, all: boolean = false): Promise<object | any[]> {
        let history = await this.getHistory(user_id)

        let dan = history?.dan

        if (dan && Object.prototype.toString.call(dan) != '[object Array]') {
            dan = [dan]
        }
        return dan ? (all ? dan : dan[0]) : undefined
    }

    /**
     * 删除 user_id 对应的存档文件
     * @param {String} user_id user_id
     */
    static async delSave(user_id: string) {
        let session = await this.get_user_token(user_id)
        if (!session) return false
        let fPath = path.join(savePath, session)
        await getFile.DelFile(path.join(fPath, 'save.json'))
        await getFile.DelFile(path.join(fPath, 'history.json'))
        await getRksRank.delUserRks(session)
        fs.rmSync(path.join(savePath, session), { recursive: true, force: true });
        this.del_user_token(user_id)
        return true
    }

    /**
     * 删除 token 对应的存档文件
     * @param {String} Token Token
     */
    static async delSaveBySessionToken(Token: string) {
        let fPath = path.join(savePath, Token)
        await getFile.DelFile(path.join(fPath, 'save.json'))
        await getFile.DelFile(path.join(fPath, 'history.json'))
        await getRksRank.delUserRks(Token)
        fs.rmSync(path.join(savePath, Token), { recursive: true, force: true });
        return true
    }

    static async banSessionToken(token: string) {
        return await redis.upsert('phigrosBanSessionToken', [{ sessionToken: token }], 'sessionToken')
    }

    static async allowSessionToken(token: string) {
        return await redis.remove('phigrosBanSessionToken', { sessionToken: token })
    }

    static async isBanSessionToken(token: string) {
        return (await redis.get('phigrosBanSessionToken', { sessionToken: token }))[0] ? true : false
    }

    static async getGod() {
        return await redis.get('phigrosBanSessionToken', { id: { $lt: -1 } }, ['sessionToken'])
    }

}
